app/Http/Controllers/Bank/NordigenController.php

```diff
@@ -17,270 +17,254 @@
-                'lang' => $lang,
-                'company' => $company,
-                'account' => $company->account,
-                'failed_reason' => "account-config-invalid",
-                "redirectUrl" => $context["redirect"] . "?action=nordigen_connect&status=failed&reason=account-config-invalid",
-            ]);
+            return $this->failed('account-config-invalid', $context, $company);
         }
 
-        if (!(Ninja::isSelfHost() || (Ninja::isHosted() && $account->isEnterprisePaidClient()))) {
-            return view('bank.nordigen.handler', [
-                'lang' => $lang,
-                'company' => $company,
-                'account' => $company->account,
-                'failed_reason' => "not-available",
-                "redirectUrl" => $context["redirect"] . "?action=nordigen_connect&status=failed&reason=not-available",
-            ]);
+        if (!(Ninja::isSelfHost() || (Ninja::isHosted() && $company->account->isEnterprisePaidClient()))) {
+            return $this->failed('not-available', $context, $company);
         }
 
         $nordigen = new Nordigen();
+        $institutions = $nordigen->getInstitutions();
 
         // show bank_selection_screen, when institution_id is not present
-        if (!array_key_exists("institution_id", $data)) {
+        if (!array_key_exists('institution_id', $data)) {
             return view('bank.nordigen.handler', [
                 'lang' => $lang,
                 'company' => $company,
                 'account' => $company->account,
-                'institutions' => $nordigen->getInstitutions(),
-                'redirectUrl' => $context["redirect"] . "?action=nordigen_connect&status=user-aborted"
+                'institutions' => $institutions,
+                'redirectUrl' => $context['redirect'] . '?action=nordigen_connect&status=user-aborted'
             ]);
         }
 
+        $institution = array_values(array_filter($institutions, function ($institution) use ($data) {
+            return $institution['id'] == $data['institution_id'];
+        }))[0];
+
+        // Renewals have an Institution ID, but bypass the history selection screen
+        // and thus lack the history setting, so we can find the Agreement ID here.
+        if (!isset($data['tx_days'])) {
+            // Query the integration so we get the correct Account ID
+            $integration = $this->findIntegrationBy('institution', $institution, $company);
+
+            // Extract the EUA ID from the expired account error
+            $match = '/End User Agreement \(EUA\) ([0-9a-f-]+) has expired/';
+            $nordigenAccount = $nordigen->getAccount($integration->nordigen_account_id);
+            $euaId = preg_replace($match, '${1}', $nordigenAccount['error']);
+
+            // Fetch the old agreement and maintain its history setting
+            $data['tx_days'] = $nordigen->getAgreement($euaId)['max_historical_days'];
+        }
```

Comment:
please explain the expected behavior here pls for me. I think this could have some difficulties, if a person has multiple accounts at one institutionId, then we can have same problems here.

a user can have multiple accounts in 2 different ways: 
1. multiple accounts per request (with the same credentials)
2. multiple accounts from different accounts (credentials)

i think you should considder rework this section to use the defaults eigher from existing accounts or by the defaults on the institution. maybe only from the institution to reduce complexity

Reply:
You're right, it's definitely not *perfect*.

The red triangle on an expired integration is still treated as a completely new connection, except the Institution ID is pre-set. The */nordigen/connect* request knows nothing about the integration, so it's a best-effort situation with the information available.

I'd read previously that only one integration/account was officially supported because of how transactions were handled/linked or something, I'm not sure... @turbo124 can you give some insight?

---

#### Options
* Since #10404, upstream disabled state could be used to filter integrations
  * More reliable, but still uses some guesswork.
* If the BankIntegration ID was in the request we could directly query it
  * Injecting the ID into token could work?
  * Separate route (e.g. `/nordigen/{bank_integration}/reconnect/{token}`) would be best
    * ...but requires rewriting half the controller
  * Injecting the ID into token could work?

**Question:** When you're on the integration list and click the expired button, where does the `institution_id` get added into the token/cache?

Reply:
bank_integration_id should be a unique id from nordigen, which is the same for each account, therefore integrations with disabled upsteam are reactivated on reconnect. (not sure if also archieved ones)
Also we have descided to dont implement 2 routes, as of to reduce complexity. after a callback (confirm) the right account is upserted.